File: Treinamento Calouros/2024/Turma 2/Contests/Simulado Primeira Fase OBI/manchas.cpp
//https://neps.academy/br/exercise/567 - Manchas de Pele

#include <bits/stdc++.h>

using namespace std;

#define endl '\n'

int n, m, ma[1005][1005], vis[1005][1005];

void dfs(int x, int y){
    if(x<1 || x>n || y<1 || y>m || ma[x][y]==0 || vis[x][y]==1) return;

    vis[x][y]=1;

    dfs(x+1, y);
    dfs(x-1, y);
    dfs(x, y+1);
    dfs(x, y-1);

}

void bfs(int x, int y){
    queue<pair<int, int> > fila;
    fila.push({x, y});

    while(!fila.empty()){
        int x = fila.front().first, y = fila.front().second;
        fila.pop();

        if(x<1 || x>n || y<1 || y>m || ma[x][y]==0 || vis[x][y]==1) continue;

        vis[x][y]=1;

        fila.push({x+1, y});
        fila.push({x-1, y});
        fila.push({x, y+1});
        fila.push({x, y-1});
    }
}

int main(){
    ios_base::sync_with_stdio(false), cin.tie(nullptr);

    cin >> n >> m;

    for(int i=1; i<=n; i++){
        for(int j=1; j<=m; j++){
            cin >> ma[i][j];
        }
    }
    int resp = 0;

    for(int i=1; i<=n; i++){
        for(int j=1; j<=m; j++){
            if(!vis[i][j] && ma[i][j]==1){
                dfs(i, j); // ou bfs(i, j)
                resp++;
            }
        }
    }
    cout << resp << endl;
    return 0;
}

File: Treinamento Calouros/2024/Turma 2/Contests/Simulado Primeira Fase OBI/manchas.py
#https://neps.academy/br/exercise/567 - Manchas de Pele

def bfs(x, y):
    fila = []
    fila.append((x, y))

    while len(fila)>0:
        (x, y) = fila[0]
        fila.pop(0)

        #lembrando que a matriz está 0-indexada, assim a posição (0, 0) é válido, mas (-1, 0) não é
        if x<0 or x>=n or y<0 or y>=m or ma[x][y]==0 or vis[x][y]==1:
            continue

        vis[x][y]=1

        fila.append((x+1, y))
        fila.append((x-1, y))
        fila.append((x, y+1))
        fila.append((x, y-1))


n, m = [int(i) for i in input().split()]

ma = []

for i in range(n):
    linha = [int(i) for i in input().split()]
    ma.append(linha)

vis = [[0 for j in range(m)] for i in range(n)] #criando matriz NxM com todas as posições iguais a zero

resp = 0

for i in range(n):
    for j in range(m):
        if vis[i][j]==0 and ma[i][j]==1:
            resp+=1
            bfs(i, j)

print(resp)
